src/sinks/influxdb_metrics.rs

```diff
@@ -0,0 +1,916 @@
+        Ok((sink, healthcheck))
+    }
+
+    fn input_type(&self) -> DataType {
+        DataType::Metric
+    }
+
+    fn sink_type(&self) -> &'static str {
+        "influxdb_metrics"
+    }
+}
+
+impl InfluxDBSvc {
+    pub fn new(config: InfluxDBConfig, cx: SinkContext) -> crate::Result<super::RouterSink> {
+        let token = config.token.clone();
+        let batch = config.batch.unwrap_or(20, 1);
+        let request = config.request.unwrap_with(&REQUEST_DEFAULTS);
+
+        let uri = format!(
+            "{}/api/v2/write?org={}&bucket={}&precision=ns",
+            config.host, config.org, config.bucket
+        )
+        .parse::<Uri>()
+        .context(super::UriParseError)?;
+        let http_service = HttpService::new(cx.resolver(), move |body: Vec<u8>| {
+            let mut builder = hyper::Request::builder();
+            builder.method(Method::POST);
+            builder.uri(uri.clone());
+
+            builder.header("Content-Type", "text/plain");
+            builder.header("Authorization", format!("Token {}", token));
+            builder.body(body).unwrap()
+        });
+
+        let influxdb_http_service = InfluxDBSvc {
+            config,
+            inner: http_service,
+        };
+
+        let sink = request
+            .batch_sink(HttpRetryLogic, influxdb_http_service, cx.acker())
+            .batched_with_min(MetricBuffer::new(), &batch);
+
+        Ok(Box::new(sink))
+    }
+
+    // https://v2.docs.influxdata.com/v2.0/api/#operation/GetHealth
+    fn healthcheck(config: InfluxDBConfig) -> crate::Result<super::Healthcheck> {
+        let uri = format!("{}/health", config.host)
+            .parse::<Uri>()
+            .context(super::UriParseError)?;
+
+        let request = hyper::Request::get(uri).body(hyper::Body::empty()).unwrap();
+
+        let https = HttpsConnector::new(4).expect("TLS initialization failed");
+        let client = hyper::Client::builder().build(https);
+
+        let healthcheck = client
+            .request(request)
+            .map_err(|err| err.into())
+            .and_then(|response| match response.status() {
+                StatusCode::OK => Ok(()),
+                other => Err(super::HealthcheckError::UnexpectedStatus { status: other }.into()),
+            });
+
+        Ok(Box::new(healthcheck))
+    }
+}
+
+impl Service<Vec<Metric>> for InfluxDBSvc {
+    type Response = HttpResponse;
+    type Error = HttpError;
+    type Future = Box<dyn Future<Item = Self::Response, Error = Self::Error> + Send + 'static>;
+
+    fn poll_ready(&mut self) -> Poll<(), Self::Error> {
+        self.inner.poll_ready()
+    }
+
+    fn call(&mut self, items: Vec<Metric>) -> Self::Future {
+        let input = encode_events(items, &self.config.namespace).join("\n");
+        let body: Vec<u8> = input.into_bytes();
+
+        self.inner.call(body)
+    }
+}
+
+fn encode_events(events: Vec<Metric>, namespace: &str) -> Vec<String> {
```

Comment:
Could this function have created `String` directly, without temporary `Vec` instance, to avoid unnecessary string copying? See the comment on `encode_key` function for more details.

Reply:
Done

Reply:
Calling `collect()` and `join()` in the end of `encode_events` function still performs an additional copying of entire batch of the data. I propose to slightly change the implementation of it to make it like this:

```rust
fn encode_events(events: Vec<Metric>, namespace: &str) -> String {
    let mut output = String::new();
    for event in events.into_iter() {
        // ...
    }
    return output;
}
```

Reply:
Yeah, I get it. Thanks

Reply:
Done